plugin/trino-loki/src/main/java/io/trino/plugin/loki/LokiMetadata.java

```diff
@@ -0,0 +1,116 @@
+import io.trino.spi.type.DoubleType;
+import io.trino.spi.type.Type;
+import io.trino.spi.type.TypeManager;
+import io.trino.spi.type.VarcharType;
+
+import java.util.List;
+import java.util.Optional;
+
+import static com.google.common.collect.ImmutableList.toImmutableList;
+import static io.trino.plugin.loki.LokiTableFunction.SCHEMA_NAME;
+import static io.trino.spi.type.TimestampWithTimeZoneType.createTimestampWithTimeZoneType;
+import static io.trino.spi.type.TypeSignature.mapType;
+import static io.trino.spi.type.VarcharType.VARCHAR;
+import static java.util.Objects.requireNonNull;
+
+public class LokiMetadata
+        implements ConnectorMetadata
+{
+    private static final String TABLE_NAME = "default";
+
+    private final Type labelsMapType;
+
+    private final LokiClient lokiClient;
+
+    static final Type TIMESTAMP_COLUMN_TYPE = createTimestampWithTimeZoneType(3);
+
+    @Inject
+    public LokiMetadata(LokiClient lokiClient, TypeManager typeManager)
+    {
+        requireNonNull(typeManager, "typeManager is null");
+        this.lokiClient = requireNonNull(lokiClient, "lokiClient is null");
+        labelsMapType = typeManager.getType(mapType(VARCHAR.getTypeSignature(), VARCHAR.getTypeSignature()));
+    }
+
+    @Override
+    public Optional<TableFunctionApplicationResult<ConnectorTableHandle>> applyTableFunction(ConnectorSession session, ConnectorTableFunctionHandle handle)
+    {
+        if (!(handle instanceof LokiTableFunction.QueryHandle queryHandle)) {
+            return Optional.empty();
+        }
+
+        LokiTableHandle tableHandle = queryHandle.getTableHandle();
+        return Optional.of(new TableFunctionApplicationResult<>(tableHandle, tableHandle.columnHandles()));
+    }
+
+    @Override
+    public ConnectorTableMetadata getTableMetadata(ConnectorSession session, ConnectorTableHandle table)
+    {
+        LokiTableHandle lokiTableHandle = (LokiTableHandle) table;
+
+        List<ColumnMetadata> columns = lokiTableHandle.columnHandles()
+                .stream()
+                .map(columnHandle -> ((LokiColumnHandle) columnHandle).columnMetadata())
+                .collect(toImmutableList());
+
+        return new ConnectorTableMetadata(new SchemaTableName(SCHEMA_NAME, TABLE_NAME), columns);
+    }
+
+    @Override
+    public ColumnMetadata getColumnMetadata(ConnectorSession session, ConnectorTableHandle tableHandle, ColumnHandle columnHandle)
+    {
+        var lokiColumnHandler = (LokiColumnHandle) columnHandle;
+        return lokiColumnHandler.columnMetadata();
+    }
+
+    public List<ColumnHandle> getColumnHandles(String query)
+    {
+        // TODO: cache result for query to avoid too many calls to Loki.
+        ImmutableList.Builder<ColumnHandle> columnsBuilder = ImmutableList.builderWithExpectedSize(3);
+        columnsBuilder.add(new LokiColumnHandle("labels", this.labelsMapType, 0));
+        columnsBuilder.add(new LokiColumnHandle("timestamp", TIMESTAMP_COLUMN_TYPE, 1));
```

Comment:
Great catch. Loki returns seconds or nanos since epoch. So it's timezone independent. The timestamp we generate does have a TZ. Should I change the column type then?

Reply:
Yes, change the column type accordingly.

Reply:
Perplexity.ai [helped](https://www.perplexity.ai/search/what-s-the-purpose-of-the-life-VV4TQCI4Qau.2zVRMSqumQ) ^^.

Reply:
I removed the lifecycle manager.